src/scheduler/server_info.c

```diff
@@ -832,6 +833,43 @@ query_server_dyn_res(server_info *sinfo)
 	return 0;
 }
 
+/**
+ * @brief	Check if there are any runjob hooks configured in the system
+ *
+ * @param	void
+ *
+ * @return	int
+ * @retval	1 if runjob hooks are configured
+ * @retval	0 if not
+ */
+static int
+check_runjob_hooks(int pbs_sd)
```

Comment:
I think it should be the server that should set the run_job_wait attribute accordingly.
But, I can understand that you are doing it in scheduler because you don't want to overwrite user-provided values in server. So to sort that out, how about the server object itself has an attribute set that says it has configured and enabled runjob hooks?
I just don't like that the way it is now, the scheduler will query hooks in every cycle and with multi-sched it will only get worse.

Reply:
I understand where you are coming from, but if you look at the performance test logs that I've posted, there's a test which specifically looks at the performance lag due to this, and it's actually not noticeable, it's just one batch request out of the thousands that sched makes every cycle, and it queries just one attribute, so the data transferred isn't much either. So I don't think it makes much of a difference.

I'd love to prevent the extra request, but I don't know if adding another server attribute is better than this.

Reply:
I think you need to query 2 attributes - enabled and type. In the test, you have only one hook. The performance will vary depending on how many runjob hooks are present in the system.
If you have a way from server to tell the scheduler that yes there is a runjob hook enabled then that's all scheduler cares about. Adding a server attribute to me is more efficient than scheduler querying it because adding/removing a hook is a rare event as compared to the number of cycles the scheduler is going to run. So once the attribute is set, it is unlikely it will change.
Using encode function is also another approach but in that case, the server will do repetitive work every scheduling cycle.

Reply:
Ok, I like the encode function idea, I didn't think about that, I'll try it out, thanks

Reply:
@arungrover about performance lag every cycle, I'm thinking of refactoring the code so that sched attributes are not queried every cycle, they will be globals which will be cached by sched, anytime it changes via qmgr, we'll update the cache. I'll do this in another PR though. So, performance won't be a problem. Please let me know if you still have any concerns.

Reply:
sounds good

Reply:
The code isn't repetitive.  The server keeps a list of every type of hook in a global.  It just needs to look to see if that global is NULL or not.  If it is kept in the scheduler, the scheduler has to ask the server for the hooks.  The server has to wrap up all the attributes as strings to send to the scheduler, send it, and then the scheduler has to unwrap it to look for a runjob hook.  I think it is much better done in the server in the encode function.

Reply:
The server also has to loop through each hook to check if it is enabled, right?

Reply:
I think i do need to go over the list of runjob hooks and check to see if any of them are enabled or not, but that shouldn't be much computation.

Reply:
I'm fine with it, especially when you said that you are going to make the scheduler cache the sched object and make it query on need basis.

Reply:
turns out, we can't know who the requester is inside the encode functions, so the encode solution won't work. I'll do the secondary server attribute solution, marking the PR as WIP for now.